Types update, SQLite fix and implementation

gionata opened the pull request:
FIX - SQLite

- not null
- primary key

UPDATES - types for SQLite

- INTEGER
-  REAL;

IMPLEMENTATION - SQLite

- unique constraints
- foreign key constraints

jycouet reviewed (approved):
Thank you for your contribution 🎉

Sorry, I had to cleanup my format/linting stuff. (So I bump deps & did a few formating things)

LGTM

jycouet commented:
@gionata you ight wanna add `ROUTE.ts` to `.prettierignore` ;)

jycouet reviewed (changes requested):
With your latest update some tests are not passing. I let you have a look.
Thank you

jycouet commented:
`bun` ? Does it has an effect ? We are using `npm` here I think.
Can't look right now, but will look in a few days.

gionata commented:
Regression: I hadn't fully considered that a SQLite database might not have the `sqlite_sequence` table when I added the _autoincrement_ check. I've now implemented a fix for this. Hope it works!

Vitest second run: My mistake! I left the automatic snapshot update enabled after the initial run.

jycouet commented on tests-examples/demo-todo-app.spec.ts:
```diff
@@ -0,0 +1,420 @@
+import { expect, test, type Page } from '@playwright/test'
+
+test.beforeEach(async ({ page }) => {
+  await page.goto('https://demo.playwright.dev/todomvc')
```
What its the test around this ?
Do we need to file at all ?

gionata replied:
We don't need this file. I was experimenting with Playwright due to some issues I had with SQLite database introspection, and these changes resulted from that. They're not relevant and can be reverted or simply not merged into the main repository.

jycouet commented on tests/e2e/landing.test.ts:
```diff
@@ -2,5 +2,7 @@ import { expect, test } from '@playwright/test'
 
 test('Landing page working, we can start 🚀', async ({ page }) => {
   await page.goto('/')
+  await page.waitForResponse((response) => response.status() === 200)
+
   expect(await page.locator('h1').textContent()).toBe('Home')
 })
```
I kinda like to have e2e test of pages directly in the `routes` folder. There is a specific reason for this change ?

jycouet commented on playwright.config.ts:
```diff
@@ -7,19 +7,25 @@ if (process.env.CI) {
+  retries: 2, // default
   // workers: 50%, // default
-  testMatch: '*/**/e2e.ts',
+  testMatch: '*/**/e2e/*.ts',
```
Maybe we could go to `'*/**/*e2e.ts',`
To have `page.e2e.ts` for example.

I'm not sure `/e2e.ts` would be matching ?!

To be checked

jycouet commented on tests/e2e/example.spec.ts:
```diff
@@ -0,0 +1,8 @@
+import { expect, test } from '@playwright/test'
+
+test('has title', async ({ page }) => {
+  await page.goto('/')
+
+  // Expect a title "to contain" a substring.
+  await expect(page).toHaveTitle(/Remult Kit/)
+})
```
We can probably regroup it with the other test ?

gionata replied:
I think the best option here is to **revert all the Playwright-related changes** I made.

Instead, it would be much more beneficial to add dedicated tests in `src/lib/cli/getEntity.test.ts` to thoroughly check the functionality implemented in `src/lib/cli/db/DbSQLite.ts`, specifically covering:

* **Types:**
    * `INTEGER`
    * `REAL`
    * `AUTOINCREMENT`
* **Constraints:**
    * `NOT NULL`
    * `UNIQUE`
    * `FOREIGN KEY`

PS: To be honest, I know nothing about Playwright.

jycouet replied:
Yes, you are right all Playwright should be reverted.

The test you suggest is a good idea, and it's a vitest one.

What's the next step ? You do these 2 things ? Or ?
Let me know ☺️

gionata replied:
I've applied the changes:

- All Playwright-related code has been reverted.  [e46761f](https://github.com/noam-honig/remult-kit/pull/29/commits/e46761fc8446df81af9c09754799057e061149c7)

- Tests for SQLite have been added. [b2bddc5](https://github.com/noam-honig/remult-kit/pull/29/commits/b2bddc52e174c2bd5cf1e7a0d866b2e530d88be8)

I've done my best with these changes, and I don't have any further plans at the moment. 

Hope it will improve the project!

gionata replied:
Thanks for merging.

There are a couple of other interesting things I'd like to implement for SQLite integration:

1. BLOB type support;
2. Creating read-only entities for VIEWs.

However, I'm having difficulty understanding how to implement them because:

1. I can't find a Uint8Array in Remult's type system;
2. I'm struggling to infer the types associated with a view's columns.

If you have any suggestions, I can try working on them in the coming weeks.
Let me know and I can open a issue or just another PR.

jycouet replied:
Maybe we can switch to a github issue and/or discord ? If not I'm worried to miss a message ;)

- For the type system, you have this option: https://remult.dev/docs/ref_valueconverter#fieldtypeindb to manage ANY db types.
- For views, I'm creating `entity` with `sqlExpression`: https://remult.dev/docs/ref_entity#sqlexpression

// Happy coding

jycouet reviewed (changes requested):
Hello hello, 
Thank you for these updates.

I would like to know what you have in mind with all these e2e tests changes ?

SQLite updates seems to work 👍

jycouet reviewed (approved):
Back to green.
Looks Good To Me ✅
